### PR TITLE
Use s3 integrator

### PR DESCRIPTION
**Pull request by arturo-seijas**
Applicable spec: <link> N/A

### Overview

Use the S3 integrator to handle the S3 configs for Indico

### Rationale

N/A

### Juju Events Changes

* Observe credentials_changed
* * Observe credentials_gone

### Module Changes

* Added s3_observer.py
* Changed charm.py
* Changed state.py

### Library Changes

* Added s3

### Checklist

- [x] The [charm style guide](https://juju.is/docs/sdk/styleguide) was applied
- [x] The [contributing guide](https://github.com/canonical/is-charms-contributing-guide) was applied
- [x] The changes are compliant with [ISD054 - Managing Charm Complexity](https://discourse.charmhub.io/t/specification-isd014-managing-charm-complexity/11619)
- [x] The documentation is generated using `src-docs`
- [x] The documentation for charmhub is updated
- [x] The PR is tagged with appropriate label (`urgent`, `trivial`, `complex`)

### CODE REVIEW

File: pyproject.toml [hunk 1]
```diff
@@ -16,7 +16,7 @@ line-length = 99
 target-version = ["py38"]
 
 [tool.coverage.report]
-fail_under = 96
+fail_under = 97
```
Comment (yanksyoon) [inline pyproject.toml:19]: 🎉

File: src/state.py [hunk 2]
```diff
@@ -5,7 +5,7 @@
 import dataclasses
 import logging
 import os
-from typing import List, Optional, Tuple
+from typing import Dict, List, Optional, Tuple
```
Comment (yanksyoon) [inline src/state.py:8]: Out of curiosity, may I know the reason is running on 20.04? (these types triggered my train of thoughts :D )

File: tests/unit/test_s3_observer.py [hunk 3]
```diff
@@ -0,0 +1,82 @@
...
+    relation_data = {
+        "bucket": "some-bucket",
+        "host": "example.s3",
+        "access-key": token_hex(16),
+        "secret-key": token_hex(16),
+    }
+    harness = Harness(ObservedCharm, meta=REQUIRER_METADATA)
+    harness.begin()
+    harness.add_relation("s3", "s3-integrator", app_data=relation_data)
+    assert len(harness.charm.events) == 1
```
Comment (yanksyoon) [inline tests/unit/test_s3_observer.py:63]: May I ask if this assertion meant to include "state updated part" in 
> a config change event is emitted and the state, updated.

?

The state seems to be tested in other unit tests :)

**Review by yanksyoon — DISMISSED** [review]
LGTM! Nitpicky comments just out of curiosity!

**Review by weiiwang01 — DISMISSED** [review]

File: src/state.py (hunk 2 above)
Comment (arturo-seijas) [reply src/state.py:8]: Iirc, the first charmed Indico had some limitations. We can now upgrade but haven't done so yet

File: tests/unit/test_s3_observer.py (hunk 3 above)
Comment (arturo-seijas) [reply tests/unit/test_s3_observer.py:63]: Updated :)

**Review by yanksyoon — DISMISSED** [review]

**Review by weiiwang01 — DISMISSED** [review]

**Review by amandahla — DISMISSED** [review]
LGTM

**Review by weiiwang01 — DISMISSED** [review]

**Review by Thanhphan1147 — DISMISSED** [review]

**Review by weiiwang01 — DISMISSED** [review]

**Review by yanksyoon — DISMISSED** [review]

File: src/state.py [hunk 4]
```diff
@@ -184,7 +214,22 @@ def from_charm(
                 if smtp_relation_data
                 else None
             )
+            s3_config = (
+                S3Config(
+                    bucket=s3_relation_data["bucket"],
+                    host=s3_relation_data["endpoint"],
```
Comment (javierdelapuente) [inline src/state.py:220]: Not sure if relevant for this PR, but in the case of AWS S3, this my be empty (same in the connection string)

**Review by javierdelapuente — DISMISSED** [review]

File: src/state.py (hunk 4 above)
Comment (arturo-seijas) [reply src/state.py:223]: Updated. Thanks!

**Review by javierdelapuente — DISMISSED** [review]

**Review by javierdelapuente — DISMISSED** [review]

**Review by gtrkiller — DISMISSED** [review]

**Review by javierdelapuente — DISMISSED** [review]

**Review by gtrkiller — DISMISSED** [review]

**Review by javierdelapuente — APPROVED** [review]

**Review by gtrkiller — APPROVED** [review]

**Pull request merged**